File: python/백준/자료구조/덱, 큐/13335 트럭.py
from collections import deque

n, w, l = map(int, input().split())
weight = list(map(int, input().split()))

bridge = deque()
time = 0
bridge_weight = 0

bridge.append([weight[0], 1])
bridge_weight += weight[0]
weight.pop(0)
time += 1

while bridge:
    for truck in bridge:
        truck[1] += 1

    if bridge[0][1] > w:
        bridge_weight -= bridge[0][0]
        bridge.popleft()
    
    if weight and bridge_weight + weight[0] <= l:
        bridge.append([weight[0], 1])
        bridge_weight += weight[0]
        weight.pop(0)
    
    time += 1

print(time)
            
    

File: cpp/백준/자료구조/덱, 큐/13335 트럭.cpp
#include <iostream>
#include <deque>
#include <vector>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, w, l;
    cin >> n >> w >> l;

    deque<int> weight;
    for (int i = 0; i < n; ++i) {
        int x;
        cin >> x;
        weight.push_back(x);
    }

    deque<pair<int, int>> bridge;
    int time = 0;
    int bridge_weight = 0;

    bridge.push_back({weight.front(), 1});
    bridge_weight += weight.front();
    weight.pop_front();
    time += 1;

    while (!bridge.empty()) {
        for (auto& truck : bridge) truck.second += 1;

        if (bridge.front().second > w) {
            bridge_weight -= bridge.front().first;
            bridge.pop_front();
        }

        if (!weight.empty() && bridge_weight + weight.front() <= l) {
            bridge.push_back({weight.front(), 1});
            bridge_weight += weight.front();
            weight.pop_front();
        }

        time += 1;
    }

    cout << time << "\n";
    return 0;
}
